src/Storage/FileSystem/HashDirectory.php

```diff
@@ -100,39 +117,27 @@ public function getHash()
             return null;
         }
 
-        $directories = $this->getDirectoryPathByHash($this->file);
-        $directories_array = explode('/', $directories);
+        $path = $this->getDirectoryPathByHash($this->file);
 
-        //create directories
-        $this->createSubDirs($directories_array[0], $directories_array[1]);
+        // Create directories
+        $this->createSubDirs($path);
 
-        return $directories;
+        return $path;
     }
 
     /**
      *  Inside $this->dir (Cache Directory), create 2 sub directories to store current cache file
      *
-     * @param $dir1 string directory
-     * @param $dir2 string directory
+     * @param string $path Relative directory path
      *
      * @throws \PageCache\PageCacheException directories not created
      */
-    private function createSubDirs($dir1, $dir2)
+    private function createSubDirs($path)
     {
-        //dir1 not exists, create both
-        if (!@is_dir($this->dir . $dir1)) {
-            mkdir($this->dir . $dir1);
-            mkdir($this->dir . $dir1 . '/' . $dir2);
-        } else {
-            //dir1 exists, create dir2
-            if (!@is_dir($this->dir . $dir1 . '/' . $dir2)) {
-                mkdir($this->dir . $dir1 . '/' . $dir2);
-            }
-        }
-        //check if directories are there
-        if (!@is_dir($this->dir . $dir1 . '/' . $dir2)) {
-            throw new PageCacheException(__CLASS__.'/'.__METHOD__.': ' . $dir1 . '/' . $dir2
-                . ' cache directory could not be created');
+        $fullPath = $this->dir.$path;
+
+        if (!\file_exists($fullPath) && !mkdir($fullPath, $this->directoryPermissions, true) && !is_dir($fullPath)) {
```

Comment:
good catch, but again i don't think it is related to our issue we are trying to solve :)

Reply:
Yes, I think that may lead the issue due concurrency. The blank page always occured on an empty cache and may be produced by unhandled error/exception during shutdown function. The race condition in directory creation is one of the probable reasons.